homeassistant/components/hanna/sensor.py

```diff
@@ -0,0 +1,292 @@
+        icon="mdi:thermometer",
+        device_class=SensorDeviceClass.TEMPERATURE,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+    ),
+    "airTemp": SensorEntityDescription(
+        key="airTemp",
+        name="Air Temperature",
+        icon="mdi:thermometer",
+        device_class=SensorDeviceClass.TEMPERATURE,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+    ),
+    "acidBase": SensorEntityDescription(
+        key="acidBase", name="pH Acid/Base Flow Rate", icon="mdi:flask"
+    ),
+    "cl": SensorEntityDescription(
+        key="cl", name="Chlorine Flow Rate", icon="mdi:chemical-weapon"
+    ),
+    "phPumpColor": SensorEntityDescription(
+        key="phPumpColor",
+        name="pH Pump Status",
+        icon="mdi:pump",
+    ),
+    "clPumpColor": SensorEntityDescription(
+        key="clPumpColor",
+        name="Chlorine Pump Status",
+        icon="mdi:pump",
+    ),
+    "StatusColor": SensorEntityDescription(
+        key="StatusColor",
+        name="System Status",
+        icon="mdi:information",
+    ),
+    "ServiceColor": SensorEntityDescription(
+        key="ServiceColor",
+        name="Service Status",
+        icon="mdi:wrench",
+    ),
+    "alarms": SensorEntityDescription(
+        key="alarms",
+        name="Alarms",
+        icon="mdi:alert",
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up Hanna sensors from a config entry."""
+    device_coordinators = entry.runtime_data["device_coordinators"]
+
+    for coordinator in device_coordinators.values():
+        if not coordinator.readings:
+            _LOGGER.warning("No data received for %s", coordinator.device_identifier)
+            continue
+
+        # Add parameter sensors
+        param_sensors = []
+        for parameter in coordinator.get_parameters():
+            if description := SENSOR_DESCRIPTIONS.get(parameter["name"]):
+                param_sensors.append(HannaParamSensor(coordinator, description))
+            else:
+                _LOGGER.warning("No sensor description found for %s", parameter["name"])
+        if param_sensors:
+            async_add_entities(param_sensors)
+
+        # Add status sensors
+        status_sensors = []
+        for sensor_name in coordinator.get_messages_value("status"):
+            if description := SENSOR_DESCRIPTIONS.get(sensor_name):
+                status_sensors.append(HannaStatusSensor(coordinator, description))
+            else:
+                _LOGGER.warning("No sensor description found for %s", sensor_name)
+        if status_sensors:
+            async_add_entities(status_sensors)
+
+        # Add alarms sensor
+        alarm_sensors = []
+        alarm_sensors.append(
+            HannaAlarmSensor(coordinator, SENSOR_DESCRIPTIONS["alarms"])
+        )
+        if alarm_sensors:
+            async_add_entities(alarm_sensors)
```

Comment:
We should collect all entities during initialisation and add them at once instead of calling async add entities multiple times